src/nypl/mod.rs

```diff
@@ -0,0 +1,204 @@
+use crate::Vec2d;
+use std::str::FromStr;
+use custom_error::custom_error;
+use std::sync::Arc;
+use crate::dezoomer::{TilesRect, Dezoomer, DezoomerInput, ZoomLevels, DezoomerError, IntoZoomLevels, DezoomerInputWithContents};
+use std::convert::TryFrom;
+use std::iter::successors;
+use std::fmt::Debug;
+use serde::export::Formatter;
+
+/// A dezoomer for NYPL images
+#[derive(Default)]
+pub struct NYPLImage;
+
+const NYPL_IMAGE_VIEW_PREFIX: &str = "https://digitalcollections.nypl.org/items/";
+const NYPL_META_PREFIX: &str = "https://access.nypl.org/image.php/";
+const NYPL_META_POSTFIX: &str = "/tiles/config.js";
+
+fn _get_image_url_from_meta_url(meta_url: &str) -> String {
+    meta_url.replace(NYPL_META_PREFIX, "")
+        .replace(NYPL_META_POSTFIX, "")
+}
+
+macro_rules! nypl_meta_format_tpl {
+    () => {
+        "https://access.nypl.org/image.php/{}/tiles/config.js"
+    }
+}
+
+impl Dezoomer for NYPLImage {
+    fn name(&self) -> &'static str { "NYPLImage" }
+    fn zoom_levels(&mut self, data: &DezoomerInput) -> Result<ZoomLevels, DezoomerError> {
+        if data.uri.starts_with(NYPL_IMAGE_VIEW_PREFIX) {
+            self.assert(data.uri.contains(NYPL_IMAGE_VIEW_PREFIX))?;
+            let image_id = data.uri.replace(NYPL_IMAGE_VIEW_PREFIX, "");
```

Comment:
Don't we want to parse the uri completely ? What if the user entered `https://digitalcollections.nypl.org/items/a14f3200-fac1-012f-f7a4-58d385a7bbd0#item-data` ?

Reply:
Currently we simply parse it without any tag or query string (like ? and #).
Maybe complex parse is reasonable? If you have strong opinion on it, I will handle more cases.

Reply:
maybe just a small regex like : `https?://digitalcollections.nypl.org/items/([a-f0-9\-]+)` ?